doc/tutorial/tutorial.md

```diff
@@ -1859,23 +1857,133 @@ Output `multiset`s are more memory-efficient than `relation`s: internally the
+    Parcel(zip, weight),
+    ZipCodes(zip, city).
+```
+
+At runtime, the client inserts a bunch of records in the `Parcel` relation,
+reads the corresponding outputs from `ParcelCity` and clears the `Parcel`
+relation since its contents is no longer needed and to avoid running out of
+memory.  This is cumbersome and expensive.
+
+A better solution is to use the DDlog **stream** abstraction, intended to
+simplify working with ephemeral data, i.e., data that gets discarded after being
+used.  A stream is a relation that only contains records added during the
+current transaction.  When the transaction completes, all previously inserted
+records as well as any records derived from them disappear without a trace and
+the memory footprint of all streams in the program drops to zero.
+
+We re-declare the `Parcel` relation as a stream (by using the `stream` keyword
+instead of `relation`).  Any rule that uses a stream yields a stream; hence
+`ParcelCity` must also be declared as a stream.
+
+```
+input relation ZipCodes(zip: u32, city: string)
+input stream Parcel(zip: u32, weight: usize)
+
+// Add the name of the destination city to each parcel.
+output stream ParcelCity(zip: u32, city: string, weight: usize)
+ParcelCity(zip, city, weight) :-
+    Parcel(zip, weight),
+    ZipCodes(zip, city).
+```
+
+No other changes to the program are required, and we can now observe how
+streams behave at runtime:
+
+```
+echo Populating the zip code table.;
+start;
+insert ZipCodes(94022, "Los Altos"),
+insert ZipCodes(94035, "Moffett Field"),
+insert ZipCodes(94039, "Mountain View"),
+insert ZipCodes(94085, "Sunnyvale"),
+commit dump_changes;
+
+echo Feeding data to the Parcel stream;
+start;
+insert Parcel(94022, 2),
+insert Parcel(94085, 6),
+commit dump_changes;
+
+# DDlog output:
+#   Populating the zip code table.
+#   Feeding data to the Parcel stream
+#   ParcelCity:
+#   ParcelCity{.zip = 94022, .city = "Los Altos", .weight = 2}: +1
+#   ParcelCity{.zip = 94085, .city = "Sunnyvale", .weight = 6}: +1
+
+echo Send more parcels!;
+start;
+insert Parcel(94039, 10),
+insert Parcel(94022, 1),
+commit dump_changes;
+
+# DDlog output:
+#   Send more parcels!
+#   ParcelCity:
+#   ParcelCity{.zip = 94022, .city = "Los Altos", .weight = 1}: +1
+#   ParcelCity{.zip = 94039, .city = "Mountain View", .weight = 10}: +1
+```
+
+Note that although the stream becomes empty at the start of every transaction,
+DDlog does not output delete records (i.e., records with negative weights) for
+streams.
+
+In general, streams can be used in rules just like normal relations.  They can
+be joined with other relations, filtered, and flat-mapped.  There are also some
+restrictions on streams.  Most importantly, streams cannot be part of a
+recursive computation.  In addition streams currently cannot be aggregated using
+the `group_by` operator, or used as one of the inputs to an antijoin.  These
+constraints may be removed in future versions of DDlog.
```

Comment:
Again, very limited. Recursion and negation are some of the cornerstones of datalog so this means streams are only available to a small set of use cases